src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+        {
+            sdkDirectory = sdkDirectory ?? s_defaultSdkDirectory;
+            return VisualBasicCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+        }
+
+        private void CSharpSourceCodeProcessing()
+        {
+            if (ExtendedOptions != null)
+            {
+                var arguments = CSharpDefaultParse(args: ExtendedOptions);
+
+                HandleCompilerErrors(arguments.Errors);
+
+                _csharpParseOptions = arguments.ParseOptions;
+                _csharpCompilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
+
+                if (WarnAsError)
+                {
+                    _csharpCompilationOptions = arguments.CompilationOptions.WithGeneralDiagnosticOption(s_generalDiagnosticOption);
+                }
+
+                _emitOptions = arguments.EmitOptions;
             }
             else
             {
-                ErrorRecord errorRecord = new ErrorRecord(
-                    new Exception(String.Format(CultureInfo.CurrentCulture, AddTypeStrings.SpecialNetVersionRequired, Language.ToString(), string.Empty)),
-                    "LANGUAGE_NOT_SUPPORTED",
-                    ErrorCategory.InvalidArgument,
-                    Language);
+                _csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
 
-                ThrowTerminatingError(errorRecord);
-                parseOptions = null;
+                if (UsingNamespace.Length > 0)
+                {
+                    _csharpCompilationOptions = _csharpCompilationOptions.WithUsings(UsingNamespace);
+                }
             }
 
-            SyntaxTree syntaxTree = CSharpSyntaxTree.ParseText(source, parseOptions);
-            var references = s_defaultAssemblies.Value;
-            if (ReferencedAssemblies.Length > 0)
+            if (String.Equals(ParameterSetName, "FromPath", StringComparison.OrdinalIgnoreCase) ||
+                String.Equals(ParameterSetName, "FromLiteralPath", StringComparison.OrdinalIgnoreCase))
             {
-                var tempReferences = new List<PortableExecutableReference>(s_autoReferencedAssemblies.Value);
-                foreach (string assembly in ReferencedAssemblies)
+                foreach (string file in _paths)
                 {
-                    if (string.IsNullOrWhiteSpace(assembly)) { continue; }
-                    string resolvedAssemblyPath = ResolveAssemblyName(assembly, true);
+                    using (var sourceFile = new FileStream(file, FileMode.Open))
+                    {
+                        var sourceText = SourceText.From(sourceFile);
+                        _syntaxTrees.Add(CSharpSyntaxTree.ParseText(sourceText, _csharpParseOptions, file));
```

Comment:
Will this handle the case that the code in those files references to symbols defined in other files?

Reply:
My understanding is yes - it is the same parser/compiler session. But I don't test this.

Reply:
Can you please verify it? `ParseText` is a static method, so it's hard to say for sure that one signle parser session is used for all those files.

Reply:
Done. Works well.
```powershell
Add-Type -Path C:\tmp\1.cs,c:\tmp\2.cs
[Test.AddType.CSharpTest2]::Add2(1,2)
3
```

1.cs
```cs
namespace Test.AddType
        {
            public class CSharpTest1
            {
                public static int Add1(int a, int b)
                {
                    return (a + b);
                }
            }
        }
```

2.cs
```cs
namespace Test.AddType
        {
            public class CSharpTest2
            {
                public static int Add2(int a, int b)
                {
                    return CSharpTest1.Add1(a, b);
                }
            }
        }
```